### ISSUE TITLE
Ajaxify render requests!

### ISSUE THREAD

**Issue by petersont** [judytuna/shortcode-site-309#13]
Currently the request to render is done with a regular hidden html form that links to a new page that's almost blank.  This puts the user in a completely new context that doesn't make sense.  They have to know to hit back, but not refresh.  It would be much cleaner to just add a message to a dom element oncsreen that says the render is in progress.  Ajax.

**Comment by judytuna** [comment]
`app/models/entry.rb` has helper definitions for that entry's image_name (different for each entry), and povserver (static). Then the actual code that requests the render from the magical ec2 machine lives in `app/views/entries/_entries.html.erb`. I suspect that breaks the MVC. But I'm not sure where I should move the "send to get rendered" code to. Maybe helpers?

**Comment by judytuna** [comment]
Also, do we want to save the image file on the Rails server (in /public i guess?), or is it okay to grab the image from the ec2 server? 

I'm also wondering what the best way to hash a given entry to an image filename so that we don't have collisions ... right now, it just uses a combination of the entry_id and user_id, so if I render something via my development server, and then Peterson renders something else (but with the same entry_id and user_id), Peterson's will overwrite mine. Low priority, but it sure makes our "production" (lol) app on heroku misleading (since we have a bunch of test renders associated to certain entry/user id combos). Here's some info on it in python: http://stackoverflow.com/questions/1303021/shortest-hash-in-python-to-name-cache-files

**Comment by petersont** [comment]
> ... I'm not sure where I should move the "send to get rendered" code to. Maybe helpers?

You're right, entry.rb doesn't seem like the home for the name of the server.  There's got to be a better place for that.  The view is not the most logical place for code that "sends to get rendered", but on the other hand it seems inconvenient to put it anywhere else because sending to get rendered means submitting an html form, and html is found in the view.

**Comment by petersont** [comment]
I see what you're saying about conflicting render jobs having the same filenames.  I didn't think of that, because I was envisioning a day when there's only one povray short code server.  What if we incorporated the povray code itself into the filename.  Like we hashed the povray code to a 6 digit number and put that along with the userid and entryid.  That would make it very unlikely that the same file be used for two different render jobs, and it would also fix a different problem: if you make an edition to your shortcode and render again, it should favor the current one, right now, there's no guarantee.

**Comment by judytuna** [comment]
Yeah! +1 for hashing in the submitted code somehow.

On Thu, Oct 6, 2011 at 1:34 AM, Peterson Trethewey
reply@reply.github.com
wrote:

**Comment by judytuna** [comment]
> >    ... I'm not sure where I should move the "send to get rendered" code to. Maybe helpers?
> 
> You're right, entry.rb doesn't seem like the home for the name of the server. There's got to be a better place for that. The view is not the most logical place for code that "sends to get rendered", but on the other hand it seems inconvenient to put it anywhere else because sending to get rendered means submitting an html form, and html is found in the view.

Actually, I tried moving pov_server and associated functions from models/entry.rb to helpers/entry_helper.rb and I could no longer call pov_server on an entry anymore. =( I don't know how entry_helper.rb works.

**Comment by petersont** [comment]
Did you try calling it not on the member?  Just call it like a function?

--pt

On Mon, Oct 17, 2011 at 1:07 AM, Judy Tuan
reply@reply.github.com
wrote:

**Comment by judytuna** [comment]
I see that you call winners (like a function, not on an Entry object) in app/views/entries/results.html.erb ... Does stuff defined in entry_helper.rb only work when called from something in an entries folder? xP

I can't access winners from rails console, either. I haven't tried messing with the views/controllers yet though.

**Comment by judytuna** [comment]
We added a "Render button" button to the form. 

This works: 

```
<form target="_blank" id="render_<%= entry.image_name %>"
    name="render_job_form"
    action="<%= entry.pov_server + 'cgi-bin/request.py' %>"
    method="post">
    <input type="hidden" name="code"
        value='<%= entry.shortcode %>' />
    <input type="hidden" name="name"
        value=<%= entry.image_name %> />
<input type="hidden" name="key"
        value=<%= entry.key %> />
    <input type="submit" value="Render button" />
</form>  
```

Pressing the "Render button" has the expected action of going to a new page (`http://ec2-107-20-100-184.compute-1.amazonaws.com/cgi-bin/request.py`) which says that it got the request to render. And it does render.

But when we add `data-remote=true` to the form tag, nothing happens when you press "Render button". No new page, no invisible render in the background. We need to learn more about `data-remote-true` =)

**Comment by petersont** [comment]
My hunch is when you set data-remote to true, rails installs a handler on the button that reads in the form and then mimics the data in the form in an xmlhttprequest.  I've tried to find a website that explicitly says that, but I can't, all I can find are websites that imply it.  Javascript then can't open the request because of something called the "same origin policy" which I am now reading about.   http://en.wikipedia.org/wiki/Same_origin_policy  makes it sound like the only reason we're able to post that form to the ec2 machine at all is because browsers accomodate legacy websites that do it.  grr.

**Comment by petersont** [comment]
My friend Maf suggested a way to make this work where we make the render request to the ec2 machine from ruby instead of from javascript.  So it still submits a form, but it submits it to a same-origin url and then we write the ruby script that receives it, and that _ruby_ script forwards the form to the ec2 machine.  That gets around the same-origin policy totally legitimately.

**Comment by petersont** [comment]
The more I think about this, the more I think that's the way to do it.  Step one is to replace the hand-rolled html form with a form made in rails like every other form in the app.  Then we get the request to render in the entry controller, and it or maybe the model submits a form using a rails library to submit a form.  I don't know how to do that, but I'll bet it's possible.

**Comment by petersont** [comment]
I believe this website says how to do it:

http://ruby-doc.org/stdlib-1.9.2/libdoc/net/http/rdoc/Net/HTTP.html

There's one line in there that does exactly what we need, we just need to put our own form into it.

**Comment by petersont** [comment]
This will also have the security security advantage that people can't see the url to our ec2 machine.

**Comment by petersont** [comment]
This is a lot easier, and I think, more convenient for the user if we do it in entry.update / entry.create, and then eliminate the render button.
